working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+static type to assign a preliminary type of `int` to the type parameter `T`.
+Then, when it proceeds to perform type inference on the function literal, the
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
+The full algorithm for type inference of invocations is expanded as follows.
+This algorithm is used for constructor invocations (including those in
+annotations, super-constructor calls, and constructor redirects), method
+invocations (including super calls), and invocations of function-typed
+expressions.
+
+1. Initial constraints: Create some constraints on type parameters by trying to
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
+   invocation target is considered the raw uninstantiated type of the class
+   enclosing the constructor declaration.)
+
+2. Downwards inference: partially solve the set of type constraints accumulated
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
+   
+3. Visit arguments (legacy): If experiment flag `inference-update-1` is not
+   enabled, recursively perform inference on all arguments to the invocation in
+   source order.  Obtain the type contexts for the recursive inference by
+   substituting the preliminary mapping (from step 2) into the corresponding
+   parameter types of the invocation target.  For each argument that is
+   recursively inferred, create additional constraints on type parameters using
+   the resulting static type.  Then proceed to step 5.
+
+4. Visit arguments (new): If experiment flag `inference-update-1` is enabled,
+   topologically sort the arguments into stages (see [topological
+   sort](#Topological-sort) below), and then for each stage, do the following:
+
+   * Recursively infer all arguments selected for this stage that are not
+     function literals (see [what constitutes a function
+     literal?](#what-constitutes-a-function-literal), in source order.  (Note
+     that an invariant of the topological sort is that non-function literals are
+     always placed in the first stage).  The logic for obtaining type contexts
+     and for creating additional constraints is the same as in step 3.
+
+   * Recursively infer all arguments selected for this stage that are function
+     literals, in source order.
+
+   * Horizontal inference: if this is not the last stage, partially solve the
+     set of type constraints accumulated so far, to produce an updated
+     preliminary mapping of type parameters to type schemas.  Otherwise proceed
+     to step 5.
+
+5. Upwards inference: solve the set of type constraints accumulated so far, to
+   produce a final mapping of type parameters to types.  Check that each type is
+   a subtype of the bound of its corresponding type parameter.
```

Comment:
I've added the term "(possibly grounded) constraint solution for a set of type variables, considering bounds" to the "terminology" section to explain how this works.  The short story is that every time we solve the constraints, if we're about to choose a solution that doesn't refer to the unknown type, then at the last minute we throw in an extra constraint to ensure that the bound is met.  The full details of how this works are tricky (especially in the case where one type parameter's bound may refer to other type parameters).  AFAICT these details are not specified anywhere, but since I'm not proposing to change them in any way, and since this proposal has already gotten quite big and detailed, I decided to draw a cut line here and not delve further into the handling of bounds.

Reply:
It is slightly dangerous to define two distinct concepts using metalanguage. So "(possibly grounded) constraint solution for a set of type variables, considering bounds" is really intended to mean "constraint solution for a set of type variables, considering bounds" (with one definition that follows), and "grounded constraint solution for a set of type variables, considering bounds" (with another definition). I often choose to use a few extra words in order to avoid introducing new, fancy meta-language mechanisms (and introducing them implicitly!), but it shouldn't be a big problem here.

Reply:
(I kept this thread unresolved, not because I'm asking for further changes, but just because I wanted to keep the previous comment visible.)